File: api-express/src/services/produto-service.js
import * as produtoDal from '../dal/produto-dal.js';

export const getAll = async (page, pageSize) => {
    return produtoDal.getAll(undefined, {
        sku: true,
        nome: true,
        foto: true,
        pctoferta: true,
        preco: true,
        descricao: true,
        estoque: true,
    }, page, pageSize);
};

export const getBySKU = async (sku) => {
    return produtoDal.getBySKU(sku, {
        sku: true,
        nome: true,
        foto: true,
        pctoferta: true,
        preco: true,
        descricao: true,
        estoque: true,
    });
};

export const getByCategoria = async (idCategoria, page, pageSize) => {
    return produtoDal.getByCategoria(idCategoria, {
        sku: true,
        nome: true,
        foto: true,
        pctoferta: true,
        preco: true,
    }, page, pageSize);
}

export const getAllOfertas = async (page, pageSize) => {
    return produtoDal.getAllOfertas({
        sku: true,
        nome: true,
        foto: true,
        pctoferta: true,
        preco: true,
        descricao: true,
        estoque: true,
    }, page, pageSize);
}

export const getAllDestaques = async (page, pageSize) => {
    return produtoDal.getAllDestaques({
        sku: true,
        nome: true,
        foto: true,
        pctoferta: true,
        preco: true,
        descricao: true,
        estoque: true,
    }, page, pageSize);
}

export const getByNome = async (nome, page, pageSize) => {
    return produtoDal.getByNome(nome, {
        sku: true,
        nome: true,
        foto: true,
        pctoferta: true,
        preco: true,
        descricao: true,
        estoque: true,
    }, page, pageSize);
};

File: api-actix/src/services/produto_service.rs
use crate::dal::produto_dal::ProdutoDal;
use crate::utils::app_message::ApiError;
use crate::db::DbPool;
use crate::models::produto::Produto;

pub struct ProdutoService;

impl ProdutoService {
    pub async fn get_all(pool: &DbPool, page: u32, page_size: u32) -> Result<Vec<Produto>, ApiError> {
        ProdutoDal::get_all(pool, "*", page, page_size).await
    }

    pub async fn get_by_sku(pool: &DbPool, sku: &String) -> Result<serde_json::Value, ApiError> {
        ProdutoDal::get_by_sku(pool, sku, "sku, produtos.nome, foto, pctOferta, preco, descricao, estoque").await
    }

    pub async fn get_by_id_categoria(pool: &DbPool, id_categoria: &String, page: u32, page_size: u32) -> Result<Vec<serde_json::Value>, ApiError> {
        ProdutoDal::get_by_id_categoria(pool, id_categoria, "sku, produtos.nome, foto, pctOferta, preco, qtdVendas", page, page_size).await
    }

    pub async fn get_all_ofertas(pool: &DbPool, page: u32, page_size: u32) -> Result<Vec<serde_json::Value>, ApiError> {
        ProdutoDal::get_all_ofertas(pool, "sku, produtos.nome, foto, pctOferta, preco", page, page_size).await
    }

    pub async fn get_all_destaques(pool: &DbPool, page: u32, page_size: u32) -> Result<Vec<serde_json::Value>, ApiError> {
        ProdutoDal::get_all_destaques(pool, "sku, produtos.nome, foto, pctOferta, preco", page, page_size).await
    }

    pub async fn get_by_nome(pool: &DbPool, nome: &String, page: u32, page_size: u32) -> Result<Vec<serde_json::Value>, ApiError> {
        ProdutoDal::get_by_nome(pool, nome, "sku, produtos.nome, foto, pctOferta, preco", page, page_size).await
    }
}